File: leetcode/2. problem/java/139.单词拆分.java
import java.time.format.TextStyle;
import java.util.List;

/*
 * @lc app=leetcode.cn id=139 lang=java
 *
 * [139] 单词拆分
 */

// @lc code=start
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        // 动态规划
        // 假设 dp[i] 表示0~i的字符串可以被分割出来
        // return extracted(s, wordDict);

        // test
        return test(s, wordDict);
    }

    private boolean test(String s, List<String> wordDict) {
        int len = s.length();
        boolean[] dp = new boolean[len + 1];
        dp[0] = true;
        for (int i = 0; i <= len; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && wordDict.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[len];
    }

    /**
     * Your runtime beats 62.97 % of java submissions
     * Your memory usage beats 31.56 % of java submissions (41.5 MB)
     * 
     * @param s
     * @param wordDict
     * @return
     */
    private boolean extracted(String s, List<String> wordDict) {
        int len = s.length();
        boolean[] dp = new boolean[len + 1];
        dp[0] = true;
        for (int i = 0; i <= len; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && wordDict.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[len];
    }
}
// @lc code=end

File: leetcode/2. problem/139.单词拆分.go
/*
 * @lc app=leetcode.cn id=139 lang=golang
 *
 * [139] 单词拆分
 */

package main

// @lc code=start
func wordBreak(s string, wordDict []string) bool {
	n := len(s)
	wordDictSet := make(map[string]bool)
	for _, w := range wordDict {
		wordDictSet[w] = true
	}
	dp := make([]bool, n+1)
	dp[0] = true
	for i := 0; i <= n; i++ {
		for j := 0; j < i; j++ {
			if dp[j] && wordDictSet[s[j:i]] {
				dp[i] = true
			}
		}
	}
	return dp[n]
}

// @lc code=end
